### PR TITLE
feat: Save control point info to URL

### LINKED ISSUE #282: Save control points to URL
## Use Case

_Please provide a use case to help us understand your request in context_

## Solution

_Please describe your ideal solution_

## Alternatives

_Please describe any alternatives you've considered, even if you've dismissed them_

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Closes #282, saving the control points to the URL!

*Estimated review size: large, 30 minutes*

This change adds the following properties to the URL:
```
controlPoints - 'cps'
controlPointsEnabled - 'cpe'
ramp - 'rmp'
```
(^ I'm open to feedback on any of these abbreviations)

These capture the current control point and ramp information, as well as whether the control points OR ramp should be visible 
 in the UI (now shown under the 'Advanced' checkbox in the volume settings editor).

Note that `rmp` is slightly redundant with the existing `lut` field in the URL; I elected to keep them separate as we expect `lut` to be a user input property, while the `ramp` and `controlPoints` are only used when capturing existing state from the user interacting with the viewer.

## Changes

- Adds `controlPoints`, `controlPointsEnabled`, and `ramp` fields to the `ViewerChannelSettings`.
- Added serializing/deserializing for control point arrays.
- Refactored the LUT initialization logic to choose between the `lut`, `ramp`, and `controlPoints` fields in the initial settings.

## Validation
- Clone and run the project locally.
- Open this link; you should see a similar channel setup to below. http://localhost:9020/viewer?url=https%253A%252F%252Fallencell.s3.amazonaws.com%252Faics%252Fnuc_morph_data%252Fdata_for_analysis%252Fbaseline_colonies%252F20200323_06_mid%252Fraw.ome.zarr%2Chttps%253A%252F%252Fallencell.s3.amazonaws.com%252Faics%252Fnuc_morph_data%252Fdata_for_analysis%252Fbaseline_colonies%252F20200323_06_mid%252Fseg.ome.zarr&mode=volumetric&mask=0&image=cell&axes=0&bb=0&bbcol=ffffff&bgcol=000000&rot=0&bright=70&dens=2.5&interp=1&reg=0%3A1%2C0%3A1%2C0%3A1&slice=0.5%2C0.5%2C0.5&lvl=35%2C140%2C255&t=0&view=Z&c0=ven%3A1%2Csen%3A0%2Cisv%3A128%2Cisa%3A1%2Cclz%3A0%2Ccza%3A1%2Ccol%3Ac3c3c3%2Ccps%3A-26.64179104477612%253A0%253Affffff%252C58.66417910447761%253A0%253A88f739%252C93.65951742627347%253A0.5412371134020619%253A88f739%252C114.85254691689009%253A0.9123711340206185%253Aff0000%252C131.26005361930297%253A0.654639175257732%253A3944f7%252C279.73880597014926%253A1%253Affffff%2Ccpe%3A1%2Crmp%3A58.66417910447761%253A136.7611940298507&c1=ven%3A0%2Csen%3A0%2Cisv%3A128%2Cisa%3A1%2Cclz%3A0%2Ccza%3A1%2Ccol%3A6fba11%2Ccps%3A0%253A0%253Affffff%252C98%253A0%253Affffff%252C115%253A1%253Affffff%252C255%253A1%253Affffff%2Ccpe%3A0%2Crmp%3A98%253A115&c2=ven%3A1%2Csen%3A0%2Cisv%3A128%2Cisa%3A1%2Cclz%3A1%2Ccza%3A1.00%2Ccol%3A8da3c0%2Ccps%3A0%253A0%253Affffff%252C88%253A1%253Affffff%252C144.24932975871315%253A0.41752577319587625%253Affffff%252C255%253A1%253Affffff%2Ccpe%3A0%2Crmp%3A226.2868632707775%253A255

image

### CODE REVIEW

File: src/aics-image-viewer/shared/utils/controlPointsToLut.ts [hunk 1]
```diff
@@ -10,66 +10,114 @@ export function controlPointsToLut(controlPoints: ControlPoint[]): Lut {
...
+ *
+ * LUT values will be determined using the following rules:
+ * - If no `lut` is provided in the `channelSettings`, a default LUT is calculated using min/max percentiles of the data.
+ * - Otherwise, `lut` will be parsed as described in `ViewerChannelSettingParams.lut`.
+ * - The `controlPoints` and `ramp` fields in the `channelSettings` will be used to override the returned "advanced mode"
+ * control points and ramp, respectively.
+ *
+ * If `controlPointsEnabled` is set to true in the `channelSettings`, the "advanced mode" control points will be applied
+ * to the volume; otherwise, the ramp will be applied.
+ */
```
Comment (ShrimpCryptid) [inline src/aics-image-viewer/shared/utils/controlPointsToLut.ts:86]: Did my best to explain what was happening in this method; open to critique!

**Review by rugeli — APPROVED** [review]
The comments are helpful! and the logic of serialization and deserialization for `cps`, `cpe` and `rmp` makes sense

**Comment by ShrimpCryptid** [comment]
> The comments are helpful! and the logic of serialization and deserialization for `cps`, `cpe` and `rmp` makes sense

Thank you, glad to hear it!

File: website/utils/url_utils.ts [hunk 2]
```diff
@@ -406,6 +440,27 @@ function parseStringRegion(region: string | undefined): PerAxis<[number, number]
...
+
+function parseControlPoints(controlPoints: string | undefined): ControlPoint[] | undefined {
+  if (!controlPoints || !CONTROL_POINTS_REGEX.test(controlPoints)) {
+    return undefined;
+  }
+  const newControlPoints = controlPoints.split(",").map((cp) => {
+    const [x, opacity, color] = cp.split(":");
+    return {
+      // TODO: Is there an expected range of values for x?
+      x: parseStringFloat(x, -Infinity, Infinity) ?? 0,
```
Comment (frasercl) [inline website/utils/url_utils.ts:455]: 0-255 is "standard," but unfortunately most anything is technically possible after a remap. This is probably as close to correct as you can get

**Review by frasercl — APPROVED** [review]
Minor comments; no blockers; LGTM!

File: website/utils/url_utils.ts [hunk 3]
```diff
@@ -28,6 +31,11 @@ const SLICE_REGEX = /^[0-9.]*,[0-9.]*,[0-9.]*$/;
  */
 const REGION_REGEX = /^([0-9.]*:[0-9.]*)(,[0-9.]*:[0-9.]*){2}$/;
 const HEX_COLOR_REGEX = /^[0-9a-fA-F]{6}$/;
+/**
+ * Matches a comma-separated list of control points, where each control point is represented
+ * by a triplet of `{x}:{opacity}:{hex color}`.
+ */
+export const CONTROL_POINTS_REGEX = /^(-?[-0-9.]*:[0-9.]*:[0-9a-fA-F]{6})(,-?[0-9.]*:[0-9.]*:[0-9a-fA-F]{6})*$/;
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply website/utils/url_utils.ts:38]: Oops, that's correct actually and I'm missing it in the second selector. This catches the case where `x` is negative.

[the remark this answers is not in the document]
Comment (frasercl) [reply website/utils/url_utils.ts:38]: Is that not the purpose of the `-?` outside the character class?

[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply website/utils/url_utils.ts:38]: Double oops, thanks for the catch!

File: src/aics-image-viewer/shared/utils/controlPointsToLut.ts [hunk 4]
```diff
@@ -10,66 +10,114 @@ export function controlPointsToLut(controlPoints: ControlPoint[]): Lut {
...
+    if (firstChar === "m") {
+      const value = parseFloat(lstr.substring(1)) / 100.0;
+      lutValues[i] = histogram.maxBin * value;
+    } else if (firstChar === "p") {
+      const value = parseFloat(lstr.substring(1)) / 100.0;
+      lutValues[i] = histogram.findBinOfPercentile(value);
+    } else {
+      lutValues[i] = parseFloat(lstr);
+    }
+  } // end for
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply src/aics-image-viewer/shared/utils/controlPointsToLut.ts]: Done!

**Pull request merged**